main.cpp

```diff
@@ -258,313 +207,199 @@ Use a service like https://www.diffchecker.com/diff to compare your output.
+template<>                                          // #6)
+struct Numeric<double>
 {
-    if( rhs == 0.0f )
+    using Type = double; //explicit
+private:
+    std::unique_ptr<Type> ud; //becomes a double
+    Numeric& powInternal(Type arg)
     {
-        std::cout << "warning: floating point division by zero!" << std::endl;
+        *ud = static_cast<Type>(std::pow(*ud, arg));
+        return *this;
     }
+public:
+    Numeric(double d) : ud( new Type(d) ) { }
     
-    *value /= rhs;
-    return *this;
-} 
-
-FloatType& FloatType::pow( float e )
-{
-    return powInternal(e);
-}
-FloatType& FloatType::pow( const IntType& e ) 
-{
-    return powInternal(e);
-}
-FloatType& FloatType::pow( const FloatType& e ) 
-{
-    return powInternal(e);
-}
-FloatType& FloatType::pow( const DoubleType& e ) 
-{
-    return powInternal(static_cast<float>(e));
-}
-FloatType& FloatType::powInternal( const float e ) 
-{
-    *value = std::pow(*value, e );
-    return *this;
-}
-FloatType& FloatType::apply(std::function<FloatType&(float&)> func)
-{
-    if(func)
+    template<typename Callable>                         // #7)
+    Numeric& apply(Callable&& f)
     {
-        return func(*value);
+        f(ud);  
+        
+        return *this;//allow for chaining
     }
-    return *this;
-}
-FloatType& FloatType::apply(void(*func)(float&))
-{
-    if(func)
+    Numeric& operator+=( const Type& rhs )
     {
-        func(*value);
+        *ud += rhs;
+        return *this;
     }
-    return *this;
-}
-
-DoubleType& DoubleType::operator+=(double rhs)
-{
-    *value += rhs;
-    return *this;
-} 
-
-DoubleType& DoubleType::operator-=(double rhs)
-{
-    *value -= rhs;
-    return *this;
-} 
-
-DoubleType& DoubleType::operator*=(double rhs)
-{
-    *value *= rhs;
-    return *this;
-} 
-
-DoubleType& DoubleType::operator/=(double rhs)
-{
-    if(fabs(rhs - 0) <= 0)
+    Numeric& operator-=( const Type& rhs )
     {
-        std::cout << "warning: floating point division by zero!" << std::endl;
+        *ud -= rhs;
+        return *this;
     }
-    *value /= rhs;
-    return *this;
-}
-DoubleType& DoubleType::pow( double e )
-{
-    return powInternal(e);
-}
-DoubleType& DoubleType::pow( const IntType& e ) 
-{
-    return powInternal(static_cast<double>(e));
-}
-DoubleType& DoubleType::pow( const FloatType& e ) 
-{
-    return powInternal(static_cast<double>(e));
-}
-DoubleType& DoubleType::pow( const DoubleType& e ) 
-{
-    return powInternal(e);
-}
-DoubleType& DoubleType::powInternal( const double e ) 
-{
-    *value = std::pow(*value, e );
-    return *this;
-}
-DoubleType& DoubleType::apply(std::function<DoubleType&(double&)> func)
-{
-    if(func)
+    Numeric& operator*=( const Type& rhs )
     {
-        return func(*value);
+        *ud *= rhs;
+        return *this;
     }
-    return *this;
-}
-DoubleType& DoubleType::apply(void(*func)(double&))
-{
-    if(func)
+    template<typename T>
+    Numeric& operator/=( const T rhs )
```

Comment:
The `Numeric<double>::operator/=` doesn't need to be templated.
That is only for the generic template class.
```
#11) now that your class is templated, you'll need to adjust your logic in your division function to handle if your input is a zero or not, based on your templated type.  
        Note: this instruction does not apply to the explicit template specialization
```
- [ ] revise